README.md
<div>
<h1>README</h1>

<div>
<h2><a id="readme-general">OASIS Open Repository: cti-taxii-client</a></h2>

<p>This GitHub public repository ( <b><a href="https://github.com/oasis-open/cti-taxii-client">https://github.com/oasis-open/cti-taxii-client</a></b> ) was created at the request of the <a href="https://www.oasis-open.org/committees/cti/">OASIS Cyber Threat Intelligence (CTI) TC</a> as an <a href="https://www.oasis-open.org/resources/open-repositories/">OASIS Open Repository</a> to support development of open source resources related to Technical Committee work.</p>

<p>While this Open Repository remains associated with the sponsor TC, its development priorities, leadership, intellectual property terms, participation rules, and other matters of governance are <a href="https://github.com/oasis-open/cti-taxii-client/blob/master/CONTRIBUTING.md#governance-distinct-from-oasis-tc-process">separate and distinct</a> from the OASIS TC Process and related policies.</p>

<p>All contributions made to this Open Repository are subject to open source license terms expressed in the <a href="https://www.oasis-open.org/sites/www.oasis-open.org/files/BSD-3-Clause.txt">BSD-3-Clause License</a>.  That license was selected as the declared <a href="https://www.oasis-open.org/resources/open-repositories/licenses">"Applicable License"</a> when the Open Repository was created.</p>

<p>As documented in <a href="https://github.com/oasis-open/cti-taxii-client/blob/master/CONTRIBUTING.md#public-participation-invited">"Public Participation Invited</a>", contributions to this OASIS Open Repository are invited from all parties, whether affiliated with OASIS or not.  Participants must have a GitHub account, but no fees or OASIS membership obligations are required.  Participation is expected to be consistent with the <a href="https://www.oasis-open.org/policies-guidelines/open-repositories">OASIS Open Repository Guidelines and Procedures</a>, the open source <a href="https://github.com/oasis-open/cti-taxii-client/blob/master/LICENSE">LICENSE</a> designated for this particular repository, and the requirement for an <a href="https://www.oasis-open.org/resources/open-repositories/cla/individual-cla">Individual Contributor License Agreement</a> that governs intellectual property.</p>

</div>

<div>
<h2><a id="purposeStatement">Statement of Purpose</a></h2>

<p>Statement of Purpose for this OASIS Open Repository (cti-taxii-client) as <a href="https://lists.oasis-open.org/archives/cti/201707/msg00000.html">proposed</a> and <a href="https://lists.oasis-open.org/archives/cti/201707/msg00001.html">approved</a> [<a href="https://issues.oasis-open.org/browse/TCADMIN-2623">bis</a>] by the TC:</p>

<p>The taxii-client under development in this GitHub repository is a Python library and command line tool for making HTTPS requests to TAXII servers in conformance with the TAXII specification.</p>

</div>

<div><h2><a id="purposeClarifications">Additions to Statement of Purpose</a></h2>

<p>Repository Maintainers may include here any clarifications &mdash; any additional sections, subsections, and paragraphs that the Maintainer(s) wish to add as descriptive text, reflecting (sub-) project status, milestones, releases, modifications to statement of purpose, etc.  The project Maintainers will create and maintain this content on behalf of the participants.</p>
</div>

<div>
<h2><a id="maintainers">Maintainers</a></h2>

<p>Open Repository <a href="https://www.oasis-open.org/resources/open-repositories/maintainers-guide">Maintainers</a> are responsible for oversight of this project's community development activities, including evaluation of GitHub <a href="https://github.com/oasis-open/cti-taxii-client/blob/master/CONTRIBUTING.md#fork-and-pull-collaboration-model">pull requests</a> and <a href="https://www.oasis-open.org/policies-guidelines/open-repositories#repositoryManagement">preserving</a> open source principles of openness and fairness. Maintainers are recognized and trusted experts who serve to implement community goals and consensus design preferences.</p>

<p>Initially, the associated TC members have designated one or more persons to serve as Maintainer(s); subsequently, participating community members may select additional or substitute Maintainers, per <a href="https://www.oasis-open.org/resources/open-repositories/maintainers-guide#additionalMaintainers">consensus agreements</a>.</p>

<p><b><a id="currentMaintainers">Current Maintainers of this Open Repository</a></b></p>

<ul>

<li><a href="mailto:gback@mitre.org">Greg Back</a>; GitHub ID: <a href="https://github.com/gtback/">https://github.com/gtback/</a>; WWW: <a href="https://www.mitre.org/">MITRE Corporation</a></li>

<li><a href="mailto:rpiazza@mitre.org">Rich Piazza</a>; GitHub ID: <a href="https://github.com/rpiazza/">https://github.com/rpiazza/</a>; WWW: <a href="https://www.mitre.org/">MITRE Corporation</a></li>

<!-- 
Greg Back, gback@mitre.org, gtback, MITRE Corporation
Rich Piazza, rpiazza@mitre.org, rpiazza, MITRE Corporation
-->

<!--<li><a href="mailto:luc.boutier@fastconnect.fr">Luc Boutier</a>; GitHub ID: <a href="https://github.com/lucboutier/">https://github.com/lucboutier/</a>; WWW: <a href="http://www.fastconnect.fr/">FastConnect SAS</a></li>  -->
</ul>

</div>

<div><h2><a id="aboutOpenRepos">About OASIS Open Repositories</a></h2>

<p><ul>
<li><a href="https://www.oasis-open.org/resources/open-repositories/">Open Repositories: Overview and Resources</a></li>
<li><a href="https://www.oasis-open.org/resources/open-repositories/faq">Frequently Asked Questions</a></li>
<li><a href="https://www.oasis-open.org/resources/open-repositories/licenses">Open Source Licenses</a></li>
<li><a href="https://www.oasis-open.org/resources/open-repositories/cla">Contributor License Agreements (CLAs)</a></li>
<li><a href="https://www.oasis-open.org/resources/open-repositories/maintainers-guide">Maintainers' Guidelines and Agreement</a></li>
</ul></p>

</div>

<div><h2><a id="feedback">Feedback</a></h2>

<p>Questions or comments about this Open Repository's activities should be composed as GitHub issues or comments. If use of an issue/comment is not possible or appropriate, questions may be directed by email to the Maintainer(s) <a href="#currentMaintainers">listed above</a>.  Please send general questions about Open Repository participation to OASIS Staff at <a href="mailto:repository-admin@oasis-open.org">repository-admin@oasis-open.org</a> and any specific CLA-related questions to <a href="mailto:repository-cla@oasis-open.org">repository-cla@oasis-open.org</a>.</p>

</div></div>


CHANGES.txt
Version 2.3.0
2021-03-12
- #95 Remove extra whitespace in Content-Type header (@2xyo)
- #94 Pagination changes to support more of the header combinations used in TAXII 2.0 (RFC7233). Provide similar as_pages method for v21 paginated call.
- #93 Fixes to ReadTheDocs not rendering Reference API documentation
- #91 Transitioned test suite from TravisCI to GitHub Workflows
- #89 Expose the `cert` parameter to all API classes in v20 and v21

Version 2.2.2
2020-09-10
- #85 Provide more details when a Server Response status code is 406. (v20, v21)
- #84 Fallback "Range" HTTP Header for pagination requests. (@teizenman) (v20)
- #81 More flexibility for the Status Endpoint validation. It no longer fails for optional successes, pendings, failures lists (@maybe-sybr) (v20, v21)

Version 2.2.1
2020-07-17
- #78 Prevent KeyError when TAXII Server Response does not include 'Content-Range' Header (v20)

Version 2.2.0
2020-07-02
- #76 drop python versions older than 3.5
- #75 fixes construction of Range Header for paginated requests RFC 7233. (@dougle)

Version 2.1.0
2020-06-09
- #68 Support for user-specified authentication including token-based authentication.
- #71 Don't log a warning when the total number of available objects is less than the per request size limit.

Version 2.0.0
2020-04-01
Base release of the TAXII Client for 2.1. Unsupported features:
- TLS support in requests
- Force connection initiation using HTTPS
Note
`import taxii2client` by default will load TAXII 2.1 classes. If this is not desired use `from taxii2client import v20`
or `from taxii2client.v20 import <class>`

Version 1.0.1
2020-04-01
- Fix packaging problem
- #60, #23 Pagination generator function available ``as_pages`` for get_objects and get_manifest requests

Version 1.0.0
2020-04-01
Base release of the TAXII Client for 2.0. Unsupported features:
- TLS support in requests
- Force connection initiation using HTTPS

Version 0.5.0
2018-11-09
- #50 Fix an issue regarding trailing slash being missing in the Endpoint url
- #51 Expose `proxies` to all Endpoint for better flexibility
- #53 Better handling of custom content. It is now stored in the `custom_properties` dictionary

Version 0.4.0
2018-10-31
- #36 Added "proxies" argument to _HTTPConnection.
- Created documentation and host on ReadTheDocs.
- #35 Handle invalid content better.
- #39 Added "_raw" to provide access to raw TAXII Responses.
- #40 Allow users to specify the user agent.

Version 0.3.1
2018-04-12
- Fix packaging problem

Version 0.3.0
2018-04-12
- #21 Re-work the accept parameters validation
- #24 Expose verify positional parameter on endpoints
- #27 Expose the accept positional parameter to allow configurable requests
- #30 Do not fail hard for optional fields of a response

Version 0.2.0
2017-11-07
- Filter keyword arguments to query params
- Add datetime formatting
- Improve README documentation
- Improve code coverage/testing

Version 0.1.0
2017-11-07
- Initial release


taxii2client/common.py
import datetime
import logging

import pytz
import requests
import requests.auth
import requests.structures
import six

from . import DEFAULT_USER_AGENT, MEDIA_TYPE_TAXII_V20, MEDIA_TYPE_TAXII_V21
from .exceptions import (
    InvalidArgumentsError, InvalidJSONError, TAXIIServiceException
)

# Module-level logger
log = logging.getLogger(__name__)


def _format_datetime(dttm):
    """Convert a datetime object into a valid STIX timestamp string.

    1. Convert to timezone-aware
    2. Convert to UTC
    3. Format in ISO format
    4. Ensure correct precision
       a. Add subsecond value if non-zero and precision not defined
    5. Add "Z"

    """

    if dttm.tzinfo is None or dttm.tzinfo.utcoffset(dttm) is None:
        # dttm is timezone-naive; assume UTC
        zoned = pytz.utc.localize(dttm)
    else:
        zoned = dttm.astimezone(pytz.utc)
    ts = zoned.strftime("%Y-%m-%dT%H:%M:%S")
    ms = zoned.strftime("%f")
    precision = getattr(dttm, "precision", None)
    if precision == "second":
        pass  # Already precise to the second
    elif precision == "millisecond":
        ts = ts + "." + ms[:3]
    elif zoned.microsecond > 0:
        ts = ts + "." + ms.rstrip("0")
    return ts + "Z"


def _ensure_datetime_to_string(maybe_dttm):
    """If maybe_dttm is a datetime instance, convert to a STIX-compliant
    string representation.  Otherwise return the value unchanged."""
    if isinstance(maybe_dttm, datetime.datetime):
        maybe_dttm = _format_datetime(maybe_dttm)
    return maybe_dttm


def _filter_kwargs_to_query_params(filter_kwargs):
    """
    Convert API keyword args to a mapping of URL query parameters.  Except for
    "added_after", all keywords are mapped to match filters, i.e. to a query
    parameter of the form "match[<kwarg>]".  "added_after" is left alone, since
    it's a special filter, as defined in the spec.

    Each value can be a single value or iterable of values.  "version" and
    "added_after" get special treatment, since they are timestamp-valued:
    datetime.datetime instances are supported and automatically converted to
    STIX-compliant strings.  Other than that, all values must be strings.  None
    values, empty lists, etc are silently ignored.

    Args:
        filter_kwargs: The filter information, as a mapping.

    Returns:
        query_params (dict): The query parameter map, mapping strings to
            strings.

    """
    query_params = {}
    for kwarg, arglist in six.iteritems(filter_kwargs):
        # If user passes an empty list, None, etc, silently skip?
        if not arglist:
            continue

        # force iterability, for the sake of code uniformity
        if not hasattr(arglist, "__iter__") or \
                isinstance(arglist, six.string_types):
            arglist = arglist,

        if kwarg == "version":
            query_params["match[version]"] = ",".join(
                _ensure_datetime_to_string(val) for val in arglist
            )

        elif kwarg == "added_after":
            if len(arglist) > 1:
                raise InvalidArgumentsError("No more than one value for filter"
                                            " 'added_after' may be given")

            query_params["added_after"] = ",".join(
                _ensure_datetime_to_string(val) for val in arglist
            )

        elif kwarg == "limit":
            query_params["limit"] = int(arglist[0])

        elif kwarg == "next":
            query_params["next"] = arglist

        else:
            query_params["match[" + kwarg + "]"] = ",".join(arglist)

    return query_params


def _to_json(resp):
    """
    Factors out some JSON parse code with error handling, to hopefully improve
    error messages.

    :param resp: A requests.Response instance
    :return: Parsed JSON.
    :raises: InvalidJSONError If JSON parsing failed.
    """
    try:
        return resp.json()
    except ValueError as e:
        # Maybe better to report the original request URL?
        six.raise_from(InvalidJSONError(
            "Invalid JSON was received from " + resp.request.url
        ), e)


def _grab_total_items_from_resource(resp):
    """Returns number of objects in bundle/envelope"""
    if isinstance(resp, requests.Response):
        resp = _to_json(resp)
    return len(resp.get("objects", []))


class TokenAuth(requests.auth.AuthBase):
    def __init__(self, key):
        self.key = key

    def __call__(self, r):
        r.headers['Authorization'] = 'Token {}'.format(self.key)
        return r


class _TAXIIEndpoint(object):
    """Contains some data and functionality common to all TAXII endpoint
    classes: a URL, connection, and ability to close the connection.  It also
    yields support in subclasses for use as context managers, to ensure
    resources are released.

    """
    def __init__(self, url, conn=None, user=None, password=None, verify=True,
                 proxies=None, version="2.0", auth=None, cert=None):
        """Create a TAXII endpoint.

        Args:
            user (str): username for authentication (optional)
            password (str): password for authentication (optional)
            verify (bool): validate the entity credentials (default: True)
            conn (_HTTPConnection): A connection to reuse (optional)
            proxies (dict): key/value pair for http/https proxy settings.
                (optional)
            version (str): The spec version this connection is meant to follow.
            cert (str or tuple): SSL client certificate default, if String,
                path to ssl client cert file (.pem). If Tuple, (‘cert’, ‘key’)
                pair. (optional)

        """
        if (conn and ((user or password) or auth)) or ((user or password) and auth):
            raise InvalidArgumentsError("Only one of a connection, username/password, or auth object may"
                                        " be provided.")
        elif conn:
            self._conn = conn
        else:
            self._conn = _HTTPConnection(user, password, verify, proxies, version=version, auth=auth, cert=cert)

        # Add trailing slash to TAXII endpoint if missing
        # https://github.com/oasis-open/cti-taxii-client/issues/50
        if url[-1] == "/":
            self.url = url
        else:
            self.url = url + "/"

    def close(self):
        self._conn.close()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
        return False


class _HTTPConnection(object):
    """This library uses the ``requests`` library, which presents a convenience
    API which hides many network details like actual connection objects.  So
    this class doesn't represent a traditional ``connection`` either.  It's a
    sort of approximation: sets of connections (or connection pools) and common
    metadata for a particular server interaction.  You can send requests to
    any hosts via the same instance; hosts/ports are not checked and new
    connection pools pop into existence as needed, but all connections are
    closed when the close() method is called.  So this is intended to be used
    for an independent self-contained interaction.

    Attributes:
        session (requests.Session): A requests session object.

    """

    def __init__(self, user=None, password=None, verify=True, proxies=None,
                 user_agent=DEFAULT_USER_AGENT, version="2.0", auth=None,
                 cert=None):
        """Create a connection session.

        Args:
            user (str): username for authentication (optional)
            password (str): password for authentication (optional)
            verify (bool): validate the entity credentials. (default: True)
            proxies (dict): key/value pair for http/https proxy settings.
                (optional)
            user_agent (str): A value to use for the User-Agent header in
                requests.  If not given, use a default value which represents
                this library.
            version (str): The spec version this connection is meant to follow.
            cert (str or tuple): SSL client certificate default, if String,
                path to ssl client cert file (.pem). If Tuple, (‘cert’, ‘key’)
                pair. (optional)
        """
        self.session = requests.Session()
        self.session.verify = verify
        # enforce that we always have a connection-default user agent.
        self.user_agent = user_agent or DEFAULT_USER_AGENT

        if user and password:
            self.session.auth = requests.auth.HTTPBasicAuth(user, password)
        elif auth:
            self.session.auth = auth

        if proxies:
            self.session.proxies.update(proxies)
        self.version = version
        if cert:
            self.session.cert = cert

    def valid_content_type(self, content_type, accept):
        """Check that the server is returning a valid Content-Type

        Args:
            content_type (str): ``Content-Type:`` header value
            accept (str): media type to include in the ``Accept:`` header.

        """
        accept_tokens = accept.replace(' ', '').split(';')
        content_type_tokens = content_type.replace(' ', '').split(';')

        if self.version == "2.0":
            return (
                    all(elem in content_type_tokens for elem in accept_tokens) and
                    (content_type_tokens[0] == 'application/vnd.oasis.taxii+json' or
                     content_type_tokens[0] == 'application/vnd.oasis.stix+json')
            )
        else:
            return (
                    all(elem in content_type_tokens for elem in accept_tokens) and
                    content_type_tokens[0] == 'application/taxii+json'
            )

    def get(self, url, headers=None, params=None):
        """Perform an HTTP GET, using the saved requests.Session and auth info.
        If "Accept" isn't one of the given headers, a default TAXII mime type is
        used.  Regardless, the response type is checked against the accept
        header value, and an exception is raised if they don't match.

        Args:
            url (str): URL to retrieve
            headers (dict): Any other headers to be added to the request.
            params: dictionary or bytes to be sent in the query string for the
                request. (optional)

        """

        merged_headers = self._merge_headers(headers)

        if self.version == "2.0":
            media_type = MEDIA_TYPE_TAXII_V20
        else:
            media_type = MEDIA_TYPE_TAXII_V21

        if "Accept" not in merged_headers:
            merged_headers["Accept"] = media_type
        accept = merged_headers["Accept"]

        resp = self.session.get(url, headers=merged_headers, params=params)

        try:
            resp.raise_for_status()
        except requests.exceptions.HTTPError as e:
            if resp.status_code == 406:
                # Provide more details about this error since its usually an import problem.
                # Import the correct version of the TAXII Client.
                logging.error(
                    "Server Response: 406 Client Error "
                    "If you are trying to contact a TAXII 2.0 Server use 'from taxii2client.v20 import X'. "
                    "If you are trying to contact a TAXII 2.1 Server use 'from taxii2client.v21 import X'"
                )
            raise e

        content_type = resp.headers["Content-Type"]
        if not self.valid_content_type(content_type=content_type, accept=accept):
            msg = (
                "Unexpected Response. Got Content-Type: '{}' for Accept: '{}'\n"
                "If you are trying to contact a TAXII 2.0 Server use 'from taxii2client.v20 import X'\n"
                "If you are trying to contact a TAXII 2.1 Server use 'from taxii2client.v21 import X'"
            )
            raise TAXIIServiceException(msg.format(content_type, accept))

        if "Range" in merged_headers and self.version == "2.0":
            return resp
        else:
            return _to_json(resp)

    def post(self, url, headers=None, params=None, **kwargs):
        """Send a JSON POST request with the given request headers, additional
        URL query parameters, and the given JSON in the request body.  The
        extra query parameters are merged with any which already exist in the
        URL.  The 'json' and 'data' parameters may not both be given.

        Args:
            url (str): URL to retrieve
            headers (dict): Any other headers to be added to the request.
            params: dictionary or bytes to be sent in the query string for the
                request. (optional)
            json: json to send in the body of the Request.  This must be a
                JSON-serializable object. (optional)
            data: raw request body data.  May be a dictionary, list of tuples,
                bytes, or file-like object to send in the body of the Request.
                (optional)
        """

        if len(kwargs) > 1:
            raise InvalidArgumentsError("Too many extra args ({} > 1)".format(
                len(kwargs)))

        if kwargs:
            kwarg = next(iter(kwargs))
            if kwarg not in ("json", "data"):
                raise InvalidArgumentsError("Invalid kwarg: " + kwarg)

        resp = self.session.post(url, headers=headers, params=params, **kwargs)
        resp.raise_for_status()
        return _to_json(resp)

    def delete(self, url, headers=None, params=None, **kwargs):
        """Perform HTTP DELETE"""
        # TODO: May need more work...
        resp = self.session.delete(url, headers=headers, params=params, **kwargs)
        resp.raise_for_status()

    def close(self):
        """Closes connections.  This object is no longer usable."""
        self.session.close()

    def _merge_headers(self, call_specific_headers):
        """
        Merge headers from different sources together.  Headers passed to the
        post/get methods have highest priority, then headers associated with
        the connection object itself have next priority.

        :param call_specific_headers: A header dict from the get/post call, or
            None (the default for those methods).
        :return: A key-case-insensitive MutableMapping object which contains
            the merged headers.  (This doesn't actually return a dict.)
        """

        # A case-insensitive mapping is necessary here so that there is
        # predictable behavior.  If a plain dict were used, you'd get keys in
        # the merged dict which differ only in case.  The requests library
        # would merge them internally, and it would be unpredictable which key
        # is chosen for the final set of headers.  Another possible approach
        # would be to upper/lower-case everything, but this seemed easier.  On
        # the other hand, I don't know if CaseInsensitiveDict is public API...?

        # First establish defaults
        merged_headers = requests.structures.CaseInsensitiveDict({
            "User-Agent": self.user_agent
        })

        # Then overlay with specifics from post/get methods
        if call_specific_headers:
            merged_headers.update(call_specific_headers)

        # Special "User-Agent" header check, to ensure one is always sent.
        # The call-specific overlay could have null'd out that header.
        if not merged_headers.get("User-Agent"):
            merged_headers["User-Agent"] = self.user_agent

        return merged_headers
